actors/src/main/scala/zio/actors/Actor.scala

```diff
@@ -2,11 +2,59 @@ package zio.actors
 
 import zio.actors.Actor.PendingMessage
 import zio.clock.Clock
+import zio.stream.ZStream
 import zio.{ Supervisor => _, _ }
 
+import scala.language.implicitConversions
+
 object Actor {
 
   private[actors] type PendingMessage[F[_], A] = (F[A], Promise[Throwable, A])
+  trait ActorResponse[R, S, +A] {
+    type ResponseType
+
+    def materialize[AA >: A](
+      state: Ref[S],
+      promise: Promise[Throwable, AA],
+      supervisor: Supervisor[R]
+    ): URIO[R with Clock, Unit]
+  }
+
+  object ActorResponse {
+    implicit def oneTime[R, S, A](response: RIO[R, (S, A)]): ActorResponse[R, S, A]                         =
+      new ActorResponse[R, S, A] {
+        override type ResponseType = RIO[R, (S, A)]
+
+        override def materialize[AA >: A](
+          state: Ref[S],
+          promise: Promise[Throwable, AA],
+          supervisor: Supervisor[R]
+        ): URIO[R with Clock, Unit] = {
+          val completer = ((s: S, a: A) => (state.set(s) *> promise.succeed(a)).ignore).tupled
+          response.foldM(
+            e =>
+              supervisor
+                .supervise[R, (S, A)](response, e)
+                .foldM(_ => promise.fail(e).ignore, completer),
+            completer
+          )
+        }
+      }
+    implicit def stream[R, S, E, I](response: ZStream[R, E, (S, I)]): ActorResponse[R, S, ZStream[R, E, I]] =
+      new ActorResponse[R, S, ZStream[R, E, I]] {
+        override type ResponseType = ZStream[R, E, (S, I)]
+
+        override def materialize[AA >: ZStream[R, E, I]](
+          state: Ref[S],
+          promise: Promise[Throwable, AA],
+          supervisor: Supervisor[R]
+        ): URIO[R with Clock, Unit] = {
+          // TODO: Supervision
+          val outputStream = response.mapM { case (s, a) => state.set(s) *> UIO(a) }
+          promise.succeed(outputStream).ignore
+        }
```

Comment:
@mtsokol @softinio @mijicd any ideas how Supervision could/should work in the streaming case?